design/003-interactive-session/003-interactive-session.md

```diff
@@ -0,0 +1,94 @@
+- Start Date: 2021-12-02
+- Status: Proposed
+
+# Run local interactive sessions
+
+## Summary
+
+Propose a new sub-group command `renku session` that is responsible for
+managing local interactive sessions for a project.
+
+## Motivation
+
+Renkulab.io provides the option to run interactive sessions of a project. Among other things the session
+contains a jupyter notebook.
+
+The motivation behind this RFC is to provide a simple renku CLI command that basically mimics this, i.e. the
+option to start an interactive session of a project locally.
+
+## Design Detail
+
+The proposed sub-commands of `renku session`, namely to run interactive sessions locally, heavily relies on
+[docker](https://www.docker.com/). Docker is one of the most popular container engines, that eases the development
+and deployment of applications, by off-loading the installation of the runtime environment dependencies of the
+application.
+
+In order to start a local interactive session, just as in case of renkulab.io, it is required to provide
+a `Dockerfile`. This is by default provided by `renku init` when creating a renku project, but if the user
+is using a custom project template it might not be present.
+
+### Proposed Commands
+
+#### renku session start
+
+A command to start an interactive session.
+
+By default, the `Dockerfile` present in the given project is going to be used for building the docker image
```

Comment:
I would say by default we look for the image on the hosted instance (if we can identify one). That would be quicker than building in most cases. We can identify the registry from the origin remote. 

There is a lot of potential for failure here so we need to think about the error handling with user-friendly suggestions for what to do next. E.g. 

* check local docker images --> none exists, check remote
* remote image lookup fails because image isn't there and no local image exists --> prompt to build one
* remote image lookup fails (private project and no token) --> suggest to login 
* remote image lookup fails, (private project with token) --> suggest image build
* image build fails, no docker --> link to docker install / documentation
* image build fails, docker working but problem in Dockerfile --> suggest a base image? or 🤷 

A related question is how do we actually identify which image we need? Simply by the commit sha? When the runtime doesn't change for many commits, we could just check which commit we need (i.e. when did the runtime dependencies change last - though we might not know all the paths in the repo that affect the runtime). Probably the safest is to go with the commit sha even if it results in too many image builds.

Reply:
by using commit hash when the runtime env doesn't change the caching mechanism (at least in case of docker) will circumvent the unnecessary rebuilds. by default the code-base itself (of the project) is mounted into the running container, so it shouldn't affect the image building at all.

Reply:
>There is a lot of potential for failure here so we need to think about the error handling with user-friendly suggestions for what to do next. E.g.
check local docker images --> none exists, check remote
remote image lookup fails because image isn't there and no local image exists --> prompt to build one
remote image lookup fails (private project and no token) --> suggest to login
remote image lookup fails, (private project with token) --> suggest image build
image build fails, no docker --> link to docker install / documentation
image build fails, docker working but problem in Dockerfile --> suggest a base image? or 🤷

we should describe this logic or do you think we should always build and never fetch from the remote registry?

Reply:
regarding the logic of image fetching/building: since now it is a plugin based solution, this is strictly up to the plugin what and how to do things. the most this RFC can do is to kind of write a suggestion how it should be handled, but honestly the plugin will do whatever it wants to do... note i'm not so sure what happens in case of notebook api when an image is defined...